qiskit/algorithms/minimum_eigensolvers/vqe.py

```diff
@@ -0,0 +1,338 @@
+                estimator_result = job.result()
+            except Exception as exc:
+                raise AlgorithmError("The primitive job to evaluate the energy failed!") from exc
+
+            values = estimator_result.values
+
+            if self.callback is not None:
+                metadata = estimator_result.metadata
+                for params, value, meta in zip(parameters, values, metadata):
+                    eval_count += 1
+                    self.callback(eval_count, params, value, meta)
+
+            energy = values[0] if len(values) == 1 else values
+
+            return energy
+
+        return evaluate_energy
+
+    def _get_evaluate_gradient(
+        self,
+        ansatz: QuantumCircuit,
+        operator: BaseOperator | PauliSumOp,
+    ) -> tuple[Callable[[np.ndarray], np.ndarray]]:
+        """Get a function handle to evaluate the gradient at given parameters for the ansatz.
+
+        Args:
+            ansatz: The ansatz preparing the quantum state.
+            operator: The operator whose energy to evaluate.
+
+        Returns:
+            A function handle to evaluate the gradient at given parameters for the ansatz.
+
+        Raises:
+            AlgorithmError: If the primitive job to evaluate the gradient fails.
+        """
+
+        def evaluate_gradient(parameters: np.ndarray) -> np.ndarray:
+            # broadcasting not required for the estimator gradients
+            try:
+                job = self.gradient.run([ansatz], [operator], [parameters])
+                gradients = job.result().gradients
+            except Exception as exc:
+                raise AlgorithmError("The primitive job to evaluate the gradient failed!") from exc
+
+            return gradients[0]
+
+        return evaluate_gradient
+
+    def _check_operator_ansatz(self, operator: BaseOperator | PauliSumOp):
+        """Check that the number of qubits of operator and ansatz match and that the ansatz is
+        parameterized.
+        """
+        if operator.num_qubits != self.ansatz.num_qubits:
+            try:
+                logger.info(
+                    "Trying to resize ansatz to match operator on %s qubits.", operator.num_qubits
+                )
+                self.ansatz.num_qubits = operator.num_qubits
+            except AttributeError as error:
+                raise AlgorithmError(
+                    "The number of qubits of the ansatz does not match the "
+                    "operator, and the ansatz does not allow setting the "
+                    "number of qubits using `num_qubits`."
+                ) from error
+
+        if self.ansatz.num_parameters == 0:
+            raise AlgorithmError("The ansatz must be parameterized, but has no free parameters.")
+
+    def _build_vqe_result(
+        self,
+        optimizer_result: OptimizerResult,
+        aux_values: ListOrDict[tuple[complex, tuple[complex, int]]],
+        eval_time: float,
+    ) -> VQEResult:
+        result = VQEResult()
```

Comment:
Is it impossible to set values through the constructor, e.g., `VQEResult(eigenvalues=..., cost_function_evals=...,)`? Setting values through setter may miss some attributes.

Reply:
I think with the current Result classes this isn't possible (without some refactoring), but it will be the natural way to set the values when we move to using dataclasses. I think the refactor to dataclasses is planned for a future PR.